### PR TITLE
[Merged by Bors] - Optimize rendering slow-down at high entity counts

### PR DESCRIPTION
**Pull request by TheRawMeatball**
# Objective

- Improve #3953

## Solution

- The very specific circumstances under which the render world is reset meant that the flush_as_invalid function could be replaced with one that had a noop as its init method.
- This removes a double-writing issue leading to greatly increased performance.

Running the reproduction code in the linked issue, this change nearly doubles the framerate.

### CODE REVIEW

**Comment by rparrett** [comment]
Neat. This is the most dramatic one of these I've seen.

image

**Comment by cart** [comment]
Wow yeah thats a big win right there. Pretty sure this is valid. I'll give this a proper think asap.

**Comment by TheRawMeatball** [comment]
Changed again to a fully custom flush method that directly uses memset, this is getting even larger gains.

**Comment by TheRawMeatball** [comment]
Currently UB, working on a fix.

**Comment by TheRawMeatball** [comment]
Reverted the memset, so this works again, but others investigating why the memset version failed via UB would be appreciated.

**Comment by TheRawMeatball** [comment]
UB fixed with help from @PROMETHIA-27

**Comment by rparrett** [comment]
b02a1d741c02174269311fa36210e15c740ba048

Screen Shot 2022-08-01 at 6 41 39 AM

File: examples/hello_world.rs [hunk 1]
```diff
@@ -1,9 +1,37 @@
-use bevy::prelude::*;
```
Comment (mockersf) [inline examples/hello_world.rs:1]: changes to this example should not be kept

File: crates/bevy_ecs/src/archetype.rs [hunk 2]
```diff
@@ -14,6 +14,7 @@ use std::{
 };
 
 #[derive(Debug, Copy, Clone, Eq, PartialEq, Hash)]
+#[repr(transparent)]
```
Comment (cart) [inline crates/bevy_ecs/src/archetype.rs:17]: Probably makes sense to add a comment to `ArchetypeId::INVALID` saying that it must be equivalent to setting all bytes of memory to `u8::MAX` (and referencing flush_and_reserve_invalid). Don't want anyone changing this here and then implicitly breaking flush_and_reserve_invalid.

**Comment by StarArawn** [comment]
I'm excited for this PR as it seems like it fixes part of the performance issues here! I'm still worried that this isn't a complete fix though. This code is still reserving entities even if they have been despawned which leads to memory leakage and possibly over time slow performance. I can think of quite a few scenarios where users would spawn high entity counts and despawn said entities. A shmup would be one such example where you would be spawning hundreds or even thousands of bullets which get depsawned.

**Comment by mbolt35** [comment]
Just checking on the general status of this PR. I noticed it was removed from the 0.8.1 release. Curious as to when we might expect this to be part of mainline. Thanks!

**Comment by hankjordan** [comment]
Noticeable performance improvement for my game as well - uses extremely high entity count, yellow is this PR, red is 0.8.0:
bench

**Pull request closed without merging**